src/dune_rules/dep_conf_eval.ml

```diff
@@ -162,6 +162,46 @@ let rec dep expander : Dep_conf.t -> _ = function
     Other
       (let* path = Expander.expand_path expander s in
        let deps = Source_deps.files path in
+       let* _, descendant_paths = Action_builder.of_memo deps in
+       let* descendant_dir_paths =
+         Action_builder.of_memo
+           (Source_tree.descendant_dir_paths (Path.drop_build_context_exn path))
+       in
+       let source_tree_path_suffix = Path.drop_build_context_exn path |> Path.source in
+       let descendant_file_paths_in_build_dir =
+         Path.Set.to_list descendant_paths |> List.map ~f:Path.drop_build_context_exn
+       in
+       List.iter descendant_dir_paths ~f:(fun descendant_dir_path ->
+         let at_least_one_build_path_descends_from_source_path =
+           List.exists
+             descendant_file_paths_in_build_dir
+             ~f:(fun descendant_file_path_in_build_dir ->
+               Path.Source.is_descendant
+                 descendant_file_path_in_build_dir
+                 ~of_:descendant_dir_path)
+         in
+         if not at_least_one_build_path_descends_from_source_path
+         then
+           User_warning.emit
```

Comment:
We have a warning module to make sure we don't emit warnings that can't be turned off.

Reply:
Ah of course, will update!